release/pkg/manager/calico/manager.go

```diff
@@ -462,22 +511,199 @@ func (r *CalicoManager) releasePrereqs() error {
+		name:  name,
+		image: component.String(),
+	}
+	r.exists, r.err = registry.ImageExists(component.ImageRef())
+	ch <- r
+}
+
+// Check that the environment has the necessary prereqs for publishing hashrelease
+func (r *CalicoManager) hashreleasePrereqs() error {
+	if r.publishHashrelease {
+		if !r.hashreleaseConfig.Valid() {
+			return fmt.Errorf("missing hashrelease server configuration")
+		}
+	}
+	images, err := pinnedversion.RetrieveComponentsToValidate(r.tmpDir)
+	if err != nil {
+		return fmt.Errorf("failed to get components to validate: %s", err)
+	}
+	if r.publishImages {
+		return r.assertImageVersions()
+	} else {
+		results := make(map[string]imageExistsResult, len(images))
+		ch := make(chan imageExistsResult)
+		for name, component := range images {
+			go imgExists(name, component, ch)
+		}
+		for range images {
+			res := <-ch
+			results[res.name] = res
+		}
+		failedImageList := []string{}
+		for name, r := range results {
+			logrus.WithFields(logrus.Fields{
+				"image":  r.image,
+				"exists": r.exists,
+			}).Info("Validating image")
+			if r.err != nil || !r.exists {
+				logrus.WithError(r.err).WithField("image", name).Error("Error checking image")
+				failedImageList = append(failedImageList, images[name].String())
+			} else {
+				logrus.WithField("image", name).Info("Image exists")
+			}
+		}
+		failedCount := len(failedImageList)
+		if failedCount > 0 {
+			return fmt.Errorf("failed to validate %d images: %s", failedCount, strings.Join(failedImageList, ", "))
+		}
+	}
+	if r.imageScanning {
+		logrus.Info("Sending images to ISS")
+		imageList := []string{}
+		for _, component := range images {
+			imageList = append(imageList, component.String())
+		}
+		imageScanner := imagescanner.New(r.imageScanningConfig)
+		err := imageScanner.Scan(imageList, r.hashrelease.Stream, false, r.tmpDir)
+		if err != nil {
+			// Error is logged and ignored as this is not considered a fatal error
+			logrus.WithError(err).Error("Failed to scan images")
+		}
 	}
+	return nil
+}
 
-	// TODO: Make sure the environment isn't dirty.
+// Check that the images exists with the correct version.
+func (r *CalicoManager) assertImageVersions() error {
+	for _, img := range images {
+		imageName := strings.TrimPrefix(img, "calico/")
+		switch img {
+		case "calico/apiserver":
+			for _, reg := range r.imageRegistries {
+				out, _ := r.runner.Run("docker", []string{"run", "--rm", fmt.Sprintf("%s/%s:%s", reg, imageName, r.calicoVersion)}, nil)
+				if !strings.Contains(out, r.calicoVersion) {
+					return fmt.Errorf("version does not match for image %s/%s:%s", reg, imageName, r.calicoVersion)
+				}
+			}
+		case "calico/cni": // cni-plugin
+			for _, reg := range r.imageRegistries {
+				for _, cmd := range []string{"calico", "calico-ipam"} {
+					out, err := r.runner.Run("docker", []string{"run", "--rm", fmt.Sprintf("%s/%s:%s", reg, imageName, r.calicoVersion), cmd, "-v"}, nil)
+					if err != nil || !strings.Contains(out, r.calicoVersion) {
```

Comment:
`err != nil` should be handled distinctly from not finding the version in the output - they are two different root causes. We def. need to log out the error if we hit it otherwise the code may fail in a confusing or non obvious way.